.github/CONTRIBUTING.md

```diff
@@ -0,0 +1,89 @@
+
+## Quality and Substance
+We do not accept changes for the sake of changes. Changes should solve real problems, not change things for your personal preferences. This does not mean changes need to be *large*, however. A spelling error is a "real problem" and is worth changing, despite being a small change. However changes such as "Changed from `for...of` to `forEach`" will likely be rejected unless some additional problem is being solved with the change.
+
+This does not mean we do not value the opinions of others, however. If you feel that a change should be made, but does not solve a specific problem (such as a refactoring change), we welcome opening a feature request for these changes. We do not claim to be infallible, and we are open to making stylistic changes to our codebases when they make sense. However changes like these must still be approved, and justified. If the changes do not provide any true substance, they will likely be rejected.
+
+Requiring changes to be approved and having substance is essential to not wasting the time of both contributors (who may spend time making changes we are not interested in) and our developers (who will have to spend time reviewing changes which ultimately get rejected).
+
+# Commits
+Besides the changes themselves, commits are the most important part of contributions. There are 2 major things to keep in mind for commits:
+
+1. [Scale/Scope](#scale-and-scope)
+2. [Messages](#messages)
+
+## Scale and Scope
+The scale and scope of a commit should be reasonable. Do not commit for every line when making multiple changes, for example. However you should not include many unrelated changes in a single commit. By limiting the scope of the commit we can ensure that if any regressions or new bugs are introduced we can easily revert those changes without the need for major refactors or reimplementations. Limiting the scale of commits also makes review of the changes easier and faster.
+
+## Messages
+Commit messages should adequately explain the changes in the commit. Messages like "Updated file.md" and "spelling error" should not be used. Nonsense messages such as "oops" or "fixed" are especially not allowed. Commit messages should, at minimum, be in the format `type: message` where `type` represents the type or scope of the changes (`chore`, `docs`, `bug-fix`, etc.) and `message` is the actual changes. Unless the word is from the codebase and starts with a capital letter (such as an exported Go struct), the `message` should be lowercase. We also recommend using both "subject" and "body" commits. This can be achieved through the git CLI by using multiple `-m`/`--message` flags. For example `git commit -m "short subject" -m "longer description of the changes"`.
```

Comment:
> the `message` should be lowercase

Unsure about having this as a requirement tbh. I personally prefer the subject being capitalized since it makes it stand out more from the description imo

The [Linux kernel](https://lore.kernel.org/asahi/) allows both lowercase and capitalized, can we also do this?

Reply:
This section was mostly inspired by https://conventionalcommits.org/, which doesn't say messages should be upper or lower case. In fact it says either acceptable, it just says to be consistent. Given that currently our existing messages seem to always be lowercase that's just the ruling I went with here, to keep things consistent. If that was an incorrect judgement I'm open to changing it. The casing itself is not that important to me, just the consistency part